src/cloudformation_cli_python_lib/identifier_utils.py

```diff
@@ -0,0 +1,79 @@
+import hashlib
+import re
+from typing import List
+
+STACK_ARN_PATTERN = "^[a-z0-9-:]*stack/[-a-z0-9A-Z]*/[-a-z0-9A-Z]*"
+
+MIN_PHYSICAL_RESOURCE_ID_LENGTH = 15
+MIN_PREFERRED_LENGTH = 17
+HASH_LENGTH = 12
+
+
+def _get_hash(client_request_token: str) -> str:
+    return hashlib.sha1(str.encode(client_request_token)).hexdigest()  # nosec
+
+
+def generate_resource_identifier(
+    stack_id_or_name: str,
+    logical_resource_id: str,
+    client_request_token: str,
+    max_length: int,
+) -> str:
+    if max_length < MIN_PHYSICAL_RESOURCE_ID_LENGTH:
+        raise Exception(
+            f"Cannot generate resource IDs shorter than\
+               {MIN_PHYSICAL_RESOURCE_ID_LENGTH} characters."
+        )
+
+    stack_name: str = stack_id_or_name
+
+    pattern = re.compile(STACK_ARN_PATTERN)
+
+    if pattern.match(stack_id_or_name):
```

Comment:
maybe just augment the regex to get a [named group](https://docs.python.org/3/howto/regex.html#non-capturing-and-named-groups). won't have to split the stack arn again to extract the name